### PR TITLE
[ShapeInference] Improve GatherND shape inference

### PR DESCRIPTION
**Pull request by mitruska**
### Details:
GatherND V5 and V8 improvements:
 - Add preserving interval dimensions
 - Add preserving labels
 - Add shape_infer function for v5::GatherND and v8::GatherND
 - Register GatherND shape infer for CPU
 - Add tests
 
### Tickets:
 - 97231

### CODE REVIEW

File: src/core/shape_inference/include/gather_nd_shape_inference.hpp [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+    }
+
+    if (data_pshape.rank().is_static() && indices_pshape.rank().is_static() &&
+        indices_pshape[indices_pshape.size() - 1].is_static()) {
+        auto batch_dims = op->get_batch_dims();
+        auto indices_tuple_length = indices_pshape[indices_pshape.size() - 1].get_length();  // last dim of indices
+
+        NODE_VALIDATION_CHECK(
+            op,
+            static_cast<int64_t>(indices_tuple_length + op->get_batch_dims()) <= data_pshape.rank().get_length(),
```
Comment (praasz) [inline src/core/shape_inference/include/gather_nd_shape_inference.hpp:40]: Is cast required ? Is not possible to make both sides same type?
If not use `cmp::le` to safe compare different types e.g:
```cpp
cmp::le(indices_tuple_length + op->get_batch_dims(), data_pshape.rank().get_length())
```

**Review by jane-intel — APPROVED** [review]

File: src/core/shape_inference/include/gather_nd_shape_inference.hpp [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+template <class TShape>
+void shape_infer(const GatherND* op, const std::vector<TShape>& input_shapes, std::vector<TShape>& output_shapes) {
+    using DimType = typename TShape::value_type;
+    output_shapes = gather_nd::gather_nd_base_shape_infer(op, input_shapes);
+
+    // If batch_dims > 1, batch dimensions are need to be fused
+    auto batch_dims = op->get_batch_dims();
+    if (batch_dims > 1 && output_shapes[0].rank().is_static()) {
+        const auto& output_base_shape = output_shapes[0];
+        std::vector<DimType> output_shape{1};
```
Comment (praasz) [inline src/core/shape_inference/include/gather_nd_shape_inference.hpp:76]: Consider to rename it `output_dims` as it is not shape class.
Consider different approach instead create vector just reuse first shape of output shapes and overwrite first dim by fused dims, drop [dim[1], dim[batch_dim]).

File: src/core/shape_inference/include/gather_nd_shape_inference.hpp [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+        for (size_t dim = 0; dim < batch_dims; ++dim) {
```
Comment (praasz) [inline src/core/shape_inference/include/gather_nd_shape_inference.hpp:77]: replace it by accumulate like
```
auto fused_dim = std::accumulate(std::next(output_base_shape.begin()),
                                 output_base_shape.begin() + batch_dims,
                                 *output_base_shape.begin(),
                                 std::multiplies<DimType>());
```                             
then this dim can be stored in
```
output_base_shape[0] = fused_dim;
```
 then 
```
output_base_shape.erase(output_base_shape.begin()+1, output_base_shape.begin()+batch_size);
```  
  can reduce creation of additional temporary variables.

**Comment by mitruska** [comment]
@praasz Thank you for the comments, as we discussed those suggestions are optional.
They will be considered within a seperate PR.

File: src/core/shape_inference/include/gather_nd_shape_inference.hpp [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+    if (data_pshape.rank().is_static() && indices_pshape.rank().is_static() &&
+        indices_pshape[indices_pshape.size() - 1].is_static()) {
+        auto batch_dims = op->get_batch_dims();
+        auto indices_tuple_length = indices_pshape[indices_pshape.size() - 1].get_length();  // last dim of indices
+
+        NODE_VALIDATION_CHECK(
+            op,
+            static_cast<int64_t>(indices_tuple_length + op->get_batch_dims()) <= data_pshape.rank().get_length(),
+            "Length of a tuple with indices must not exceed a rank of data tensor excluding batch dimensions.");
+        int64_t slice_length = data_pshape.rank().get_length() - indices_tuple_length - batch_dims;
```
[the remark this answers is not in the document]
Comment (mitruska) [reply src/core/shape_inference/include/gather_nd_shape_inference.hpp:42]: Changed to auto, applied in PR: https://github.com/openvinotoolkit/openvino/pull/15416

File: src/core/shape_inference/include/gather_nd_shape_inference.hpp [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+        int64_t output_indices_length = indices_pshape.rank().get_length() - batch_dims - 1;
+        auto output_rank = output_indices_length + slice_length;
+
+        using DimType = typename TShape::value_type;
+        std::vector<DimType> output_shape(output_rank + batch_dims);
```
[the remark this answers is not in the document]
Comment (mitruska) [reply src/core/shape_inference/include/gather_nd_shape_inference.hpp:47]: Applied in PR: https://github.com/openvinotoolkit/openvino/pull/15416

File: src/core/shape_inference/include/gather_nd_shape_inference.hpp [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+                                  "Batch dimensions of data and indices must be the same.");
+        }
+        for (int64_t dim = 0; dim < output_indices_length; ++dim) {
+            output_shape[batch_dims + dim] = indices_pshape[batch_dims + dim];
+        }
+        for (int64_t dim = 0; dim < slice_length; ++dim) {
+            output_shape[batch_dims + output_indices_length + dim] =
+                data_pshape[batch_dims + indices_tuple_length + dim];
+        }
+        return std::vector<TShape>{TShape(output_shape)};
```
[the remark this answers is not in the document]
Comment (mitruska) [reply src/core/shape_inference/include/gather_nd_shape_inference.hpp:60]: Applied in PR: https://github.com/openvinotoolkit/openvino/pull/15416

File: src/core/shape_inference/include/gather_nd_shape_inference.hpp (hunk 1 above)
Comment (mitruska) [reply src/core/shape_inference/include/gather_nd_shape_inference.hpp:40]: Updated to use `cmp::le`, applied in PR: https://github.com/openvinotoolkit/openvino/pull/15416

File: src/core/shape_inference/include/gather_nd_shape_inference.hpp (hunk 3 above)
Comment (mitruska) [reply src/core/shape_inference/include/gather_nd_shape_inference.hpp:77]: Updated to use accumulate and applied in PR: https://github.com/openvinotoolkit/openvino/pull/15416
Currently PartialShape class doesn't implement `erase`, can be considered as a future improvements.

**Comment by mitruska** [comment]
@praasz All of the comments are resolved and applied in the PR: https://github.com/openvinotoolkit/openvino/pull/15416

**Pull request merged**